tracing-core/src/field.rs

```diff
@@ -864,66 +875,77 @@ impl<'a> ValueSet<'a> {
     /// Visits all the fields in this `ValueSet` with the provided [visitor].
     ///
     /// [visitor]: Visit
-    pub fn record(&self, visitor: &mut dyn Visit) {
-        let my_callsite = self.callsite();
-        for (field, value) in self.values {
-            if field.callsite() != my_callsite {
-                continue;
-            }
-            if let Some(value) = value {
-                value.record(field, visitor);
-            }
-        }
+    pub fn record(&self, visitor: &mut dyn ValuableVisit) {
+        self.values.visit(visitor);
     }
 
     /// Returns `true` if this `ValueSet` contains a value for the given `Field`.
     pub(crate) fn contains(&self, field: &Field) -> bool {
-        field.callsite() == self.callsite()
-            && self
-                .values
-                .iter()
-                .any(|(key, val)| *key == field && val.is_some())
+        field.callsite() == self.callsite() && {
+            let mut visitor = KeyContainsVisitor {
+                field: NamedField::new(field.name()),
+                res: false,
+            };
+            self.values.visit(&mut visitor);
+            visitor.res
+        }
     }
 
     /// Returns true if this `ValueSet` contains _no_ values.
     pub(crate) fn is_empty(&self) -> bool {
-        let my_callsite = self.callsite();
-        self.values
-            .iter()
-            .all(|(key, val)| val.is_none() || key.callsite() != my_callsite)
+        let mut visitor = IsEmptyVisitor { res: false };
+        self.values.visit(&mut visitor);
+        visitor.res
```

Comment:
Wait isn't checking `named_values.is_empty()` O(1) though? Or maybe I'm misunderstanding what you mean. Is it because of the recursion that it's O(n)?